src/librustc_span/lib.rs

```diff
@@ -1089,14 +1089,12 @@ impl SourceFile {
             hasher.finish::<u128>()
         };
         let end_pos = start_pos.to_usize() + src.len();
-        if end_pos > u32::max_value() as usize {
-            return Err(OffsetOverflowError);
-        }
+        assert!(end_pos <= u32::max_value() as usize);
```

Comment:
Is there any benefit/are we able to type `end_pos` as a `u32` if we're going to assert that it fits in a `u32` or, for indexing purposes (or something else) is it important that it's a `usize` that can fit in a `u32`?

Reply:
The type we use [here](https://github.com/rust-lang/rust/pull/69266/files#diff-ec48da82b30ef2b77de7ef58ea62ccbcR137) is 32bit so it this needs to fit within it. The reason that's an `AtomicU32` is because `AtomicU64` is not available on 32 bit platforms.

Reply:
Oh, sorry, I think my run-on-sentence question was confusing :-)

I guess I'm wondering why `end_pos` is typed as a `usize` instead of a `u32`.

I guess the `usize` could be smaller than 32bit? Or is this just because rust forces us to use `usize` when doing memory stuff?

Reply:
Oh, it's because `end_pos as u32 <= u32::max_value()` is always `true`. Even on 64 bit environments because the value will be truncated. Doing it this way means it's essentially a `no-op` on 32 bit platforms and it works correctly on 64 bit platforms. 

The very cautious thing to do would be to use `checked_add` here as well but this should already have been checked down [here](https://github.com/rust-lang/rust/pull/69266/files/cc4375b5110e7395efbc803c60dbc2ef83bb2604#diff-ec48da82b30ef2b77de7ef58ea62ccbcR217-R218) where we use `checked_add`

Reply:
Oh hmmm yeah maybe this was the wrong line to put the comment on. I wasn't thinking
```rs
end_pos as u32 <= u32::max_value()
```
as much as
```rs
let end_pos: u32 = start_pos.to_u32() + src.len().to_u32();
```
or something. But if I'm reading you correctly it sounds like `to_u32()` would truncate the value instead of panicking saying, "This can't be safely cast to `u32`!"?

Reply:
There's no `usize::to_u32()` :)

If you do `u32::try_from()` you get `Option<u32>` which is `None` if the value is too large. If you do `foo as u32` that never fails but will truncate data if it's too large.

There's no convenience method because `u32::try_from(val).expect("too large")` is simple enough.

Reply:
❗️Ah! Thank you :-). Okay so I think I'm all caught up on this :-). Just to check that I've got it all:
1. First we make sure that when we instantiate a new `SourceFile`, it never has an `end_pos` greater than `u32::max_value()` (which, I assume (⚠️) means that the `SourceFile`'s total space is never more than `u32::max_value()`).
1. Then we can use `AtomicU32` to represent the used space of the `SourceFile`.
1. Then, when `SourceMap` is allocating space for `SourceFile`s, we ensure that the `size` we're allocating fits in 32bits and we can than leverage `AtomicU32` to update that value safely?

Reply:
Yes, exactly!